lightningd/datastore.c

```diff
@@ -289,6 +289,51 @@ static struct command_result *json_deldatastore(struct command *cmd,
 	return command_success(cmd, response);
 }
 
+static struct command_result *json_datastoreusage(struct command *cmd,
+						 const char *buffer,
+						 const jsmntok_t *obj UNNEEDED,
+						 const jsmntok_t *params)
+{
+	struct json_stream *response;
+	const char **k, **key;
+	struct db_stmt *stmt;
+	const u8 *data;
+	u64 gen, total_bytes = 0;
+
+	if (!param(cmd, buffer, params,
+		   p_opt("key", param_list_or_string, &key),
+		   NULL))
+		return command_param_failed();
+
+	// We ignore an empty key string or key array.
+	if (key && *key[0] == '\0')
+		key = NULL;
+
+	response = json_stream_success(cmd);
+	json_object_start(response, "datastoreusage");
+	json_add_string(response, "key", datastore_key_fmt(tmpctx, key));
+
+	for (stmt = wallet_datastore_first(cmd, cmd->ld->wallet, key,
+					   &k, &data, &gen);
+	     stmt;
+	     stmt = wallet_datastore_next(cmd, key, stmt,
+	     				  &k, &data, &gen)) {
+
+		u64 self_bytes = tal_bytelen(data);
+		for (size_t i = 0; i < tal_count(k); i++) {
+			self_bytes += strlen(k[i]);
```

Comment:
This isn't quite right, since the actual space in the db includes the tal_count(k)-1 separators.  Note that tal_count(k) is always >=1 though.

Reply:
Thanks for your review!
Uh yeah I guess I was still asleep when I did this yesterday. I forgot to double check the db schema. The PK of the table is the key as a `BLOB` with key entries separated by a string terminator `\0`. I forgot to count these and added a fix `self_bytes += tal_count(k) - 1;`